build/org.eclipse.cdt.build.gcc.core/src/org/eclipse/cdt/build/gcc/core/ClangToolChain.java

```diff
@@ -23,12 +31,131 @@
+		if (pathToToolChain == null) {
+			return envVars;
+		}
+
+		List<IEnvironmentVariable> envVarsNew = new ArrayList<>();
+		// envVars can be null if the user defined tc did not specify any env variables.
+		if (envVars != null) {
+			envVarsNew.addAll(Arrays.asList(envVars));
+		}
+
+		for (String type : new String[] { CC, CXX }) {
+			if (isAbsent(type, envVarsNew)) {
+				IPath mapped = mapCompilerPath(IPath.fromPath(pathToToolChain), type);
+				if (mapped != null) {
+					envVarsNew.add(new EnvironmentVariable(type, mapped.toString()));
+				}
+			}
+		}
+		return envVarsNew.toArray(new EnvironmentVariable[0]);
+	}
+
+	private static boolean isAbsent(String name, List<IEnvironmentVariable> envVarsNew) {
+		return envVarsNew.stream().noneMatch(ev -> {
+			if (Platform.OS_WIN32.equals(Platform.getOS())) {
+				return ev.getName().equalsIgnoreCase(name);
+			} else {
+				return ev.getName().equals(name);
+			}
+		});
+	}
+
+	/**
+	 * Given a compiler path and a desiredType ("CC" or "CXX"), return the correct path for that type.
+	 * If the path already matches the desired type, return it as-is.
+	 *
+	 * @param pathToToolChain Path to the filename of the CC or CXX compiler. It may or may not have an extension.
+	 *   Supports clang toolchains, with optional version suffix.
+	 *   Does not support cc or c++. gcc and g++ are supported in {@link GCCToolChain}.
+	 *
+	 * @param desiredType The desired compiler type to convert to.
+	 * When {@link #CC}, clang++ or clang++-version is converted to clang or clang-version.
+	 * When {@link #CXX}, clang or clang-version is converted to clang++ or clang++-version.
+	 *
+	 * For example:
+	 *   clang, clang.exe, clang-18, clang-18.exe
+	 * converts to:
+	 *   clang++, clang++.exe, clang++-18, clang++-18.exe
+	 *
+	 * @return mapped compiler path to the filename of the desired compiler.
+	 *   Returns null if desiredType is not recognised. Returns null if originalPath is to an unsupported compiler.
+	 */
+	private static IPath mapCompilerPath(IPath pathToToolChain, String desiredType) {
+		if (!CC.equals(desiredType) && !CXX.equals(desiredType)) {
+			// Invalid desired compiler type
+			return null;
+		}
+
+		// Work with just the filename part, not the entire path.
+		String filename = pathToToolChain.lastSegment();
+
+		Matcher cxxMatcher = CXX_PATTERN.matcher(filename);
+		Matcher ccMatcher = CC_PATTERN.matcher(filename);
+		boolean isCXX = cxxMatcher.matches();
+		boolean isCC = ccMatcher.matches();
+		if (isCC || isCXX) {
+			// Compiler is either clang or clang++
+			String mappedFilename;
+			if (isCXX && CC.equals(desiredType)) {
+				// Compiler is CPP type and desired type is C
+				String version = cxxMatcher.group(1) != null ? "-" + cxxMatcher.group(1) : EMPTY; //$NON-NLS-1$
+				String ext = cxxMatcher.group(2) != null ? cxxMatcher.group(2) : EMPTY;
+				mappedFilename = "clang" + version + ext; //$NON-NLS-1$
```

Comment:
It seems we are splitting the filename into multiple parts using regex groups and then re-combining them here. Could we simplify by using a group that captures the entire filename instead? The existing regex groups describing the optional parts would be nested inside this top-level group. Similarly for `GCCToolChain`.

Reply:
Hi @betamaxbandit, in your `clang` tool regex you have: `clang(?:-(\\d+))?(\\.exe)?`

You have capturing groups for the optional major version (group 1) and the optional .exe suffix (group 2).

You could use nested groups as follows: `(clang(?:-(\\d+))?(\\.exe)?)`

In that case, group 1 would be the full file name, group 2 would be the major version and group 3 would be the optional .exe suffix. So you would not need to re-assemble the full file name from its constituent parts. In fact, the inner groups could all be non-capturing: `(clang(?:-\\d+)?(?:\\.exe)?)`

I am suggesting that you could eliminate some code that way.

Reply:
Hi @betamaxbandit, do you have time to look at this simplification at present? If not, I suggest we merge your existing change for the time being. It's a good improvement.